mir: split up `mnkLiteral`

zerbina opened the pull request:
## Summary

Split up `mnkLiteral` into its constituent kinds and replace the
`PNode`-based data representation for literals with a data-oriented
one. This removes the last `ref` type usage from within the `MirNode`
record-case. In addition, comparing numeric values becomes a single
integer comparison.

## Details

`mnkLiteral` is split into `mnkNilLit`, `mnkIntLit`, `mnkUIntLit`,
`mnkFloatLit`, and `mnkAstLit`. Same as with the CGIR, numeric type
width is not part of the syntax (it's not needed).

The `MirNode` variant for the numeric literal nodes stores a
`NumberId`, which is a `LitId` uniquely identifying a bit-pattern
(stored as `BiggestInt` in `MirEnv.numbers`).

A shared storage is used because:
* it's easier to implement (one table instead of three; no special
  handling for floats is needed)
* less memory usage (value storage is shared between the three numeric
  literal kinds)
* only one ID type is needed, instead of three
* reinterpreting bit-patterns is made slightly easier

Literal AST (identified by the new `AstId` type) doesn't use a table,
only a `Store`. Their sparse usage doesn't warrant the implementation
complexity / run-time overhead of hashing/comparing the trees (though
`treetabs` could be used).

### Translation

* for integer and float literals, `mirgen` performs the same
  translation that `cgirgen` previously did in `translateLit`
* so that the code generators don't need to be adjusted, the CGIR still
  stores the numeric and AST values directly in `CgNode`
* to prevent excessive parameter passing in `cgirgen`, a pointer to the
  `MirEnv` is stored in `TranslateCl`. Access to the environment is
  needed for looking up the values for `NumberId` and `AstId`
* `compat.translate` now requires a `MirEnv` argument, for looking up
  the numerical values
* `cgirgen.translateLit` is obsolete and thus removed
* raw field symbols are transported to the code generator via a
  `mnkAstLit`/`cnkAstLit` (a `mnkLiteral` node was previously used)

### VM backend

* the `MirEnv.numbers` table is re-used for the `PackedEnv`. This
  allows for also re-using the `NumberId`s (they can be treated
  directly as a `LitId`)
* the serialization logic is adjusted to the new MIR syntax

### Tests

* the `mnkLiteral` usage in `tdatatables.nim` is replaced
* `tcompilerapi.nim` passed untyped float values to
  `nimeval.callRoutine`, which no longer works, since `constDataToMir`
  always expects typed literal nodes now. The test is changed to pass
  typed `nkFloatLit` nodes

zerbina committed:
mir: split up `mnkLiteral`

Each kind of literal now uses its own node kind: `mnkNilLit`,
`mnkIntLit`, `mnkUIntLit`, `mnkFloatLit`, and `mnkAstLit`.

The numerical values all share the same underlying storage, with the
reason being that:
* it easier to implement (one table instead of three; no special
  handling for floats is needed)
* less memory usage (values are shared across the three literal
  kinds)
* one ID type instead of three
* reinterpreting bit-patterns is made slightly easier

Literal AST doesn't use a table, only a `Store`. Their sparse usage
doesn't warrant the implementation complexity / run-time overhead of
hashing/comparing the trees (though `treetabs` could be used).

zerbina committed:
update `mnkLiteral` usage

What it's replaced with is context-dependent.

zerbina committed:
mirgen: update translation of literals

`mirgen` now has to perform the translation previously done by
`cgirgen`.

zerbina committed:
utils: update the pretty-printer

* literal values are renderer as their ID if no environment is
  available
* unsigned values are rendered with an `'u` suffix, to distinguish
  them from signed ones
* ast values are rendered as just `<Ast>` for now

zerbina committed:
injecthooks: remove `boolLit` usage

Using a `PNode` for a literal integer value no longer works.

zerbina committed:
update `mnkLiteral` usage in `ConstrTree`

zerbina committed:
cgirgen: update translation of literals

The CGIR keeps storing the values directly, so that the code generator
don't have to be adjusted for now. `translateLit`, which is obsolete
now, is removed.

zerbina committed:
mirgen: adjust field symbol boxing

For using raw field symbols in `asm`/`.emit` interpolation, the raw
node needs to be passed through to the code generator (at least at the
moment). `mnkLiteral` was used for this so far, but it's gone now, so
`mnkAstLit` is used in its place.

zerbina committed:
tests: update `--expandArc`-using tests

Enums (which includes booleans) aren't rendered as their high-level
field names anymore.

zerbina committed:
update `PackedEnv` production

The `numbers` BiTable of `MirEnv` is re-used for the `PackedEnv`, with
the `NumberId`s from `MirNode` being turned directly into a `LitId`
for serialization.

zerbina committed:
tests: update `tdatatables.nim`

`mnkLiteral` is replaced with the new node kinds.

zerbina committed:
tests: fix `tcompilerapi.nim`

The float literals passed as the arguments had no type, which was
previously okay, but now it isn't. Both have the correct type assigned,
fixing the test failure.

saem commented on compiler/mir/mirenv.nim:
```diff
@@ -43,8 +43,13 @@ type
       ## includes both normal globals and threadvars
     procedures*: SymbolTable[ProcedureId, PSym]
 
+    numbers*: BiTable[BiggestInt]
```
Would it be too much to have an `mnkEnumField` kind and a store for the field values?

zerbina replied:
I think it would be too much, yeah. The existence of enum fields is irrelevant to the mid-end and code generators -- they only care about the numeric values.

Nonetheless, if you want, I can update the pretty-printer to render enum values as their field names again.

saem replied:
> Nonetheless, if you want, I can update the pretty-printer to render enum values as their field names again.

That would be nice if not too much trouble, just makes tracing and debugging easier.

zerbina replied:
I've updated the MIR pretty-printing of literals to produce results that match the previous ones (the PR description is updated already).

saem replied:
Thanks

saem reviewed (approved):
Bit of a bummer to lose the enum field names, but I also understand it might cause a lot of issues putting that mapping in.

zerbina committed:
utils: partially restore old rendering of numeric values

Rendering of numeric values now takes the type into account again,
making it behave similarly to the old rendering.

zerbina committed:
Revert "tests: update `--expandArc`-using tests"

This reverts commit e8ab99b4576af3a4495586371a10b46e37e43ec1.

saem commented:
/merge
